packages/storage-plus/src/path.rs

```diff
@@ -0,0 +1,70 @@
+use serde::de::DeserializeOwned;
+use serde::Serialize;
+use std::marker::PhantomData;
+
+use crate::helpers::{may_deserialize, must_deserialize, nested_namespaces_with_key};
+use cosmwasm_std::{to_vec, StdError, StdResult, Storage};
+
+pub struct Path<T>
+where
+    T: Serialize + DeserializeOwned,
+{
+    /// all namespaces prefixes and concatenated with the key
+    pub(crate) storage_key: Vec<u8>,
+    // see https://doc.rust-lang.org/std/marker/struct.PhantomData.html#unused-type-parameters for why this is needed
+    data: PhantomData<T>,
+}
+
+impl<T> Path<T>
+where
+    T: Serialize + DeserializeOwned,
+{
+    pub fn new(namespace: &[u8], keys: &[&[u8]]) -> Self {
+        let l = keys.len();
+        // FIXME: make this more efficient
+        let storage_key = nested_namespaces_with_key(&[namespace], &keys[0..l - 1], keys[l - 1]);
```

Comment:
If I understand correctly, this just builds a "long name" to access a given key under a given namespace. That is, there's no concept of a directory (or a hierarchy of maps), but one map with with "composable paths", which are built a way that avoids name collision issues.

This can be enough for the use cases. But it's here, I would say, where it differs from a fs approach, which implements a specific "directory" data type (a kind of map, in the end). And, perhaps more importantly, puts these "maps" in a hierarchical order.

Reply:
Yeah, we currently allow `(namespace, key)`, `(namespace, key1, key2)` and `(namespace, key1, key2, key3)` through the various `PrimaryKey` implementations. When a map is defined, all data has the same depth, but we can define one map of "depth 2" and another of "depth 4".